e2e/key/key.go

```diff
@@ -0,0 +1,235 @@
+			consoleOps: []e2e.SingularityConsoleOp{
+				e2e.ConsoleSendLine("0"),
+				e2e.ConsoleSendLine("e2etests"),
+			},
+		},
+	}
+
+	prepCmd := func(exportPath string, secret, armor bool) []string {
+		cmd := []string{"export"}
+
+		if armor {
+			cmd = append(cmd, "--armor")
+		}
+		if secret {
+			cmd = append(cmd, "--secret")
+		}
+
+		cmd = append(cmd, exportPath)
+
+		return cmd
+	}
+
+	for _, tt := range tests {
+		c.env.RunSingularity(
+			t,
+			e2e.WithPrivileges(false),
+			e2e.WithCommand("key"),
+			e2e.WithArgs(prepCmd(tt.exportPath, tt.secret, tt.armor)...),
+			e2e.ConsoleRun(tt.consoleOps...),
+			e2e.ExpectExit(0),
+		)
+	}
+}
+
+// singularityKeyImport will export a private, and public (binary and ASCII) key.
+func (c *ctx) singularityKeyImport(t *testing.T) {
+	tests := []struct {
+		name       string
+		exportPath string
+		consoleOps []e2e.SingularityConsoleOp
+	}{
+		{
+			name:       "import public binary",
+			exportPath: c.publicExportPath,
+			consoleOps: []e2e.SingularityConsoleOp{},
+		},
+		{
+			name:       "import private binary",
+			exportPath: c.privateExportPath,
+			consoleOps: []e2e.SingularityConsoleOp{
+				e2e.ConsoleSendLine("e2etests"),
+				e2e.ConsoleSendLine("e2etests"),
+				e2e.ConsoleSendLine("e2etests"),
+			},
+		},
+		{
+			name:       "import public ascii",
+			exportPath: c.publicExportASCIIPath,
+			consoleOps: []e2e.SingularityConsoleOp{},
+		},
+		{
+			name:       "import private ascii",
+			exportPath: c.privateExportASCIIPath,
+			consoleOps: []e2e.SingularityConsoleOp{
+				e2e.ConsoleSendLine("e2etests"),
+				e2e.ConsoleSendLine("e2etests"),
+				e2e.ConsoleSendLine("e2etests"),
+			},
+		},
+	}
+
+	prepCmd := func(exportPath string) []string {
+		return []string{"import", exportPath}
+	}
+
+	for _, tt := range tests {
+		c.singularityResetKeyring(t) // Remove the tmp keyring before each import
+		c.env.RunSingularity(
+			t,
+			e2e.WithPrivileges(false),
+			e2e.WithCommand("key"),
+			e2e.WithArgs(prepCmd(tt.exportPath)...),
+			e2e.ConsoleRun(tt.consoleOps...),
+			e2e.ExpectExit(0),
+		)
+	}
+	//t.Run("singularityKeyExport", c.singularityKeyExport)
+}
+
+func (c *ctx) singularityResetKeyring(t *testing.T) {
+	// TODO: run this as non-root
+	if err := os.RemoveAll(c.keyRing); err != nil {
+		t.Fatalf("unable to remove tmp keyring directory: %s", err)
+	}
+}
+
+// RunE2ETests is the main func to trigger the test suite
+func RunE2ETests(env e2e.TestEnv) func(*testing.T) {
+	c := &ctx{
+		env:                    env,
+		publicExportPath:       filepath.Join(env.TestDir, "public_key.asc"),
+		publicExportASCIIPath:  filepath.Join(env.TestDir, "public_ascii_key.asc"),
+		privateExportPath:      filepath.Join(env.TestDir, "private_key.asc"),
+		privateExportASCIIPath: filepath.Join(env.TestDir, "private_ascii_key.asc"),
+		keyRing:                filepath.Join(env.TestDir, "sypgp-test-keyring"),
+	}
+
+	if err := os.Setenv("SINGULARITY_SYPGPDIR", c.keyRing); err != nil {
```

Comment:
this should be passed as part of the environment of the command to be run, not as part of the environment in the running test.

Reply:
I'm setting `SINGULARITY_SYPGPDIR` to a tmp directory so it can easily be removed after each `import`. I'm not specifically trying to test `SINGULARITY_SYPGPDIR`, but its helpful for testing other `key` commands.

Reply:
I think @mem is pointing at another potential problem: setting `SINGULARITY_SYPGPDIR` as it is done here will prevent us from safely run the test in parallel since the caller (i.e., this function) may end up set that environment variable and start various e2e test at the same time. We have the same problem with the image cache (I am actually working on this) but we would need code modifications to solve this problem in a clean way. I would therefore personally suggest to create an issue on the topic and just put a comment here saying that this is a temporary solution.

To be more precise, my idea to solve this type of problems is to extend the `singularityCmd` structure in `singularitycmd.go` to include both a `cacheDir` and a `syPGPDir` elements. If not defined (empty string), during the execution of `RunSingularity`, we simply create temporary directories and use them for PGP and the image cache.